openhcl/page_pool_alloc/src/lib.rs

```diff
@@ -269,7 +582,66 @@ impl PagePool {
         }
     }
 
-    // TODO: save method and restore
+    /// Validate that all allocations have been restored. This should be called
+    /// after all devices have been restored.
+    ///
+    /// `leak_unrestored` controls what to do if a matching allocation was not
+    /// restored. If true, the allocation is marked as leaked and the function
+    /// returns Ok. If false, the function returns an error if any are
+    /// unmatched.
+    ///
+    /// Unmatched allocations are always logged via a `tracing::warn!` log.
+    pub fn validate_restore(&self, leak_unrestored: bool) -> anyhow::Result<()> {
+        let mut inner = self.inner.lock();
+        let mut unrestored_allocation = false;
+
+        // Mark unrestored allocations as leaked.
+        for state in inner.state.iter_mut() {
+            if let State::AllocatedPendingRestore {
+                base_pfn,
+                pfn_bias,
+                size_pages,
+                device_id,
+                tag,
+            } = state
+            {
+                tracing::warn!(
+                    base_pfn = *base_pfn,
+                    pfn_bias = *pfn_bias,
+                    size_pages = *size_pages,
+                    device_id = *device_id,
+                    tag = tag.as_str(),
+                    "unrestored allocation"
+                );
+
+                if leak_unrestored {
+                    *state = State::Leaked {
+                        base_pfn: *base_pfn,
+                        pfn_bias: *pfn_bias,
+                        size_pages: *size_pages,
+                        device_id: *device_id,
+                        tag: tag.clone(),
+                    };
+                }
+
+                unrestored_allocation = true;
+            }
+        }
+
+        // Mark unrestored device ids as leaked.
+        for device_id in inner.device_ids.iter_mut() {
+            if let DeviceId::PendingRestore(name) = device_id {
+                tracing::warn!(device_id = name.as_str(), "unrestored device id");
+                *device_id = DeviceId::Leaked(name.clone());
+            }
+        }
+
+        if unrestored_allocation && !leak_unrestored {
+            Err(anyhow::anyhow!("unrestored allocations found for pool"))
```

Comment:
nit: I'd probably let the caller just decide what to do by returning a special error type `HasUnrestoredAllocations` or something, rather than push a bool in that means "ignore the error".